test/tests/performance/pbs_history_cleanup_quasihang.py

```diff
@@ -0,0 +1,106 @@
+# For more information, contact Altair at www.altair.com.
+#
+# This file is part of the PBS Professional ("PBS Pro") software.
+#
+# Open Source License Information:
+#
+# PBS Pro is free software. You can redistribute it and/or modify it under the
+# terms of the GNU Affero General Public License as published by the Free
+# Software Foundation, either version 3 of the License, or (at your option) any
+# later version.
+#
+# PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+# details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# The PBS Pro software is licensed under the terms of the GNU Affero General
+# Public License agreement ("AGPL"), except where a separate commercial license
+# agreement for PBS Pro version 14 or later has been executed in writing with
+# Altair.
+#
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software - under
+# a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+from tests.performance import *
+
+
+class HistoryCleanupQuasihang(TestPerformance):
+    """
+    This test suite aims at testing the quasihang caused by a lot of jobs
+    in the history.
+    Without the fix, the server takes a lot of time to respond to a client.
+    With the fix, the amount of time is significantly reduced.
+    """
+
+    def setUp(self):
+        TestPerformance.setUp(self)
+
+        a = {'job_history_enable': 'True', "job_history_duration": "10:00:00"}
+        self.server.manager(MGR_CMD_SET, SERVER, a)
+
+        a = {ATTR_rescavail + '.ncpus': 100}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
+
+    @timeout(60000)
+    def test_time_for_stat_during_history_cleanup(self):
+        """
+        This test case submits 50k very short jobs so that the job history
+        has a lot of jobs in it.
+        After submitting the jobs, the job_history_duration is reduced so
+        that the server could start purging the job histiry.
+
+        Another job is submitted and stat-ed. The test case then finds the
+        amount of time the server takes to respond.
+
+        The test case is not designed to pass/fail on builds with/without
+        the fix.
+        """
+        test = []
+        test += ['echo test\n']
+        # Submit a lot of jobs.
+        for i in range(0, 10):
+            for k in range(0, 10):
+                for l in range(0, 500):
```

Comment:
codacy will complain about the index variables that aren't used.  Say for _ in range():

also, why the 3 loops?  Why not for _ in range(0, 50000)

Reply:
On my system, if I do not sleep for a second, the system becomes unresponsive for a while. It has been my experience that the testbed systems are even slower. That explains one extra for loop.
The other for loop ensures that the just submitted 5000 jobs have finished.

Reply:
I have a test which submits 5000 jobs without a problem.  I just turn off scheduling first.  If you know you are going to submit 5000 jobs, there is no reason to continuously run scheduling cycles during the submission process.  Turn off scheduling beforehand and turn it back on when you are done with the submissions.  You won't need the sleep.

Reply:
I am trying to send jobs to the scheduler so that they finish running sooner. If I turn off the scheduling and wait for 50k jobs to get over it takes more time. I believe I would like to stick with the current logic. Similarly, I tried running the test with 500 ncpus, it takes more than 5 hours, while 100 ncpus it gets over within 4 hours. It may be only on my system. I have the logs to support what I just said.

Reply:
OK, still running less scheduling cycles is better than more.  It means the server is more responsive.  How about turning off and on scheduling in the 3rd loop?  This means you'll run one cycle per 500 jobs.

Reply:
until you insist, I would like to keep it as it is. I don't think that turning off/on the scheduling will drastically reduce the amount of time the test case will take to finish.

Reply:
I think it will make a dent.  Each scheduling cycle is a series of about 10 stat calls to the server.  Each time the server is servicing one of those stat calls, it isn't accepting a job.  What is your aversion to running less cycles?  In any case, it's your test and it's just as correct one way or the other.  I'll leave it up to you.

Reply:
I do not avert to that. My concern is the time it takes for the test to complete. I just tried doing what you suggested and increased the number of ncpus to 500. The first 5000 jobs itself takes more time than earlier. That is the reason I want to stick with this logic.

Reply:
I'm surprised it takes more time, but if it does then your logic is better.

Reply:
yes, it takes more time - at least on my system. 
So, shall I consider this as an approval?